src/middlewared/middlewared/main.py

```diff
@@ -1106,14 +1116,37 @@ def register_hook(self, name, method, *, inline=False, order=0, raise_error=Fals
 
         self.__hooks[name].append({
             'method': method,
+            'blockable': blockable,
             'inline': inline,
             'order': order,
             'raise_error': raise_error,
             'sync': sync,
         })
         self.__hooks[name] = sorted(self.__hooks[name], key=lambda hook: hook['order'])
 
+    @contextlib.contextmanager
+    def block_hooks(self, *names):
+        for name in names:
+            if not self.__hooks[name]:
+                raise RuntimeError(f'Hook {name!r} does not exist')
+
+            if not self.__hooks[name][0]['blockable']:
```

Comment:
That's exactly what I want: so when someone blocks a hook, they walk through all its instances and evaluate the consequences.

Reply:
Your logic has a chance to not work. If I block `core.on_connect, 'check_permission'` (for example) and `check_permission` is the first method in the `self.__hooks[0][name][0]` list then your logic won't catch it.

Reply:
Probably needs to be something like
```
if not_all_blocked := [i['method'].__qualname__ for i in self.__hooks[0]['blockable']]:
    raise RuntimeError(f'Hook {name!r} cannot be blocked because these methods aren't blockable: {', '.join(not_all_blocked)}')

Reply:
@yocalebo there is an additional check when adding a hook to ensure all hooks with the same name have the same `blockable` value https://github.com/truenas/middleware/pull/8807/files#diff-4ef37d706a216131586ad2e34ac622509009a88fcd0a315f98973fb29c3cfc74R1099

Reply:
Then why have the check `if not self.__hooks[name][0]['blockable']` then?

Reply:
To check if we are trying to block an unblockable hook :)

Reply:
I don't understand 😄 If I do `register_hook('core.on_connect', 'check_permission', blockable=True)` then it's going to fail to start at line 1099. However, if I change all `register_hook` instances to be `blockable=True` then when will line 1133 every work?

Reply:
If you do `register_hook('core.on_connect', 'check_permission', blockable=True)` in one file and `register_hook('core.on_connect', 'do_other_thing', blockable=False)` (default) in other file then one of these two will fail, preventing the middleware from starting up. That will ensure that if we want to block a hook, we'll set `blockable=True` for all its registrations explicitly.

Reply:
I'm truly sorry, I still don't see the need for both of these checks. (I'm probably being an idiot and completely missing it but here is what I see)

First iteration of the for loop on line 1099 will see this `register_hook('core.on_connect', 'check_permission', blockable=True)` will continue on, right? The next iteration in the for loop will see `register_hook('core.on_connect', 'blah_blah', blockable=False) and raise a `RuntimeWarning` so middlewared is stopped and you have to go fix the other instance. You'll get blocked at `reigster_hook` line 1099 until you change all hooks being registered, right? So when will line 1133 ever be true after this?

Reply:
1133 will be true if, for example, a new developer comes, sees how we block `'core.on_connect'` and decides to block, e.g. `'pool.post_import'` just by using `with self.middleware.block_hooks('pool.post_import'):` without first going through all their registrations and marking them as blockable

Reply:
Wow...I have no idea why my brain couldn't see that....I guess it's time for me to step away from the computer for a bit 🤯 thanks